### PR TITLE
Allow people to create lists at the common layer, instead of having to go through the C# and VB layers.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-compiler @jaredpar (public API changes).

This change moves list creation functionality down to the common layer.  Currently to create syntax lists one must go through CSharp.SyntaxFactory or VB.SyntaxFactory APIs.  this is necessary even though the code in both layers are the same, and can all be available at teh common layer.  This means that IDE features (like https://github.com/dotnet/roslyn/pull/20517) must have language specific code, even if they're performing language neutral operations.

This split was necessary in the early days of Roslyn (when we didn't have a common syntax model).  But now that we do, we can share much more.

**Comment by CyrusNajmabadi** [comment]
Note: i haven't done SeparateSyntaxList yet.  Those factory methods are interesting in that some of them synthesize the separator token.  As the separator token is language specific those methods cannot be moved down.

**Comment by CyrusNajmabadi** [comment]
@jcouv @gafter Could you take a look?

**Comment by CyrusNajmabadi** [comment]
Agreed.  I'd like @gafter and @jaredpar 's thoughts on this.  Though we would take this through the public API meetings.  My choices are:

1. Normal constructors (this PR).
2. Static factory methods on the respective types.  i.e. SyntaxList.Create(...)
3. A common location that can have static factory methods.  i.e. CommonSyntaxFactory.CreateList(...)

Personally i like '1'.

File: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt [hunk 1]
```diff
@@ -1,5 +1,6 @@
...
 *REMOVED*Microsoft.CodeAnalysis.Emit.EmitOptions.EmitOptions(bool metadataOnly = false, Microsoft.CodeAnalysis.Emit.DebugInformationFormat debugInformationFormat = (Microsoft.CodeAnalysis.Emit.DebugInformationFormat)0, string pdbFilePath = null, string outputNameOverride = null, int fileAlignment = 0, ulong baseAddress = 0, bool highEntropyVirtualAddressSpace = false, Microsoft.CodeAnalysis.SubsystemVersion subsystemVersion = default(Microsoft.CodeAnalysis.SubsystemVersion), string runtimeMetadataVersion = null, bool tolerateErrors = false, bool includePrivateMembers = false, System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.Emit.InstrumentationKind> instrumentationKinds = default(System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.Emit.InstrumentationKind>)) -> void
+Microsoft.CodeAnalysis.CommandLineArguments.DisplayLangVersions.get -> bool
```
Comment (jcouv) [inline src/Compilers/Core/Portable/PublicAPI.Unshipped.txt]: >CommandLineArguments.DisplayLangVersions [](start = 23, length = 40)

I don't understand why `CommandLineArguments.DisplayLangVersions` is showing up as a diff in this PR.

File: src/Compilers/Core/Portable/Syntax/SyntaxList`1.cs [hunk 2]
```diff
@@ -23,6 +24,42 @@ internal SyntaxList(SyntaxNode node)
...
+        }
+
+        private static SyntaxNode CreateNode(IEnumerable<TNode> nodes)
+        {
+            if (nodes == null)
+            {
+                return null;
+            }
+
+            var collection = nodes as ICollection<TNode>;
```
Comment (jcouv) [inline src/Compilers/Core/Portable/Syntax/SyntaxList`1.cs]: Consider inlining with a pattern: `nodes is ICollection<TNode> collection ? ....`

File: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt (hunk 1 above)
Comment (jcouv) [inline src/Compilers/Core/Portable/PublicAPI.Unshipped.txt]: Never mind, it was just re-ordered.

File: src/Compilers/Core/Portable/Syntax/SyntaxList`1.cs [hunk 3]
```diff
@@ -23,6 +24,42 @@ internal SyntaxList(SyntaxNode node)
...
+        public SyntaxList(IEnumerable<TNode> nodes)
+            : this(CreateNode(nodes))
+        {
+        }
+
+        private static SyntaxNode CreateNode(IEnumerable<TNode> nodes)
+        {
+            if (nodes == null)
+            {
+                return null;
```
Comment (jcouv) [inline src/Compilers/Core/Portable/Syntax/SyntaxList`1.cs]: Why do we throw ArgumentNullException in SyntaxNodeOrTokenList.CreateNode, but return `null` here?

Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/SyntaxList`1.cs]: Good question.  But I'm just trying to preserve existing behavior.

File: src/Compilers/Core/Portable/Syntax/SyntaxTokenList.cs [hunk 4]
```diff
@@ -31,14 +32,65 @@ internal SyntaxTokenList(SyntaxNode parent, GreenNode tokenOrList, int position,
...
+        }
+
+        private static GreenNode CreateNode(SyntaxToken[] tokens)
+        {
+            if (tokens == null)
+            {
+                return null;
+            }
+
+            // TODO: we could remove the unnecessary builder allocations here and go directly
```
Comment (jcouv) [inline src/Compilers/Core/Portable/Syntax/SyntaxTokenList.cs]: No TODOs please. File an issue instead.
Same question on `null` case above (should it throw?).

File: src/Compilers/Core/Portable/Syntax/SyntaxList`1.cs [hunk 5]
```diff
@@ -23,6 +24,42 @@ internal SyntaxList(SyntaxNode node)
...
+            var collection = nodes as ICollection<TNode>;
+            var builder = (collection != null) ? new SyntaxListBuilder<TNode>(collection.Count) : SyntaxListBuilder<TNode>.Create();
+
+            foreach (TNode node in nodes)
+            {
+                builder.Add(node);
+            }
+
+            return builder.ToList().Node;
+        }
```
Comment (jcouv) [inline src/Compilers/Core/Portable/Syntax/SyntaxList`1.cs]: Please add explicit compiler-level tests for new public APIs.

File: src/Compilers/Core/Portable/Syntax/SyntaxTokenList.cs (hunk 4 above)
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/SyntaxTokenList.cs]: Keeping existing behavior.

File: src/Compilers/Core/Portable/Syntax/SyntaxList`1.cs (hunk 5 above)
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/SyntaxList`1.cs]: Sure.

**Pull request merged**